hudi-client/hudi-client-common/src/test/java/org/apache/hudi/table/action/TestCleanPlanner.java

```diff
@@ -0,0 +1,221 @@
+    when(mockHoodieTable.getHoodieView()).thenReturn(mockFsView);
+    SyncableFileSystemView sliceView = mock(SyncableFileSystemView.class);
+    when(mockHoodieTable.getSliceView()).thenReturn(sliceView);
+    when(sliceView.getPendingCompactionOperations()).thenReturn(Stream.empty());
+    when(sliceView.getPendingLogCompactionOperations()).thenReturn(Stream.empty());
+    HoodieTableMetaClient metaClient = mock(HoodieTableMetaClient.class);
+    when(mockHoodieTable.getMetaClient()).thenReturn(metaClient);
+    HoodieTableConfig tableConfig = new HoodieTableConfig();
+    when(metaClient.getTableConfig()).thenReturn(tableConfig);
+    HoodieTimeline mockCompletedCommitsTimeline = mock(HoodieTimeline.class);
+    when(mockCompletedCommitsTimeline.countInstants()).thenReturn(100);
+    when(mockHoodieTable.getCompletedCommitsTimeline()).thenReturn(mockCompletedCommitsTimeline);
+  }
+
+  @ParameterizedTest
+  @MethodSource("keepLatestByHoursArgs")
+  void keepLatestByHours(String earliestInstant, List<HoodieFileGroup> allFileGroups, List<Pair<String, Option<byte[]>>> savepoints,
+                         List<HoodieFileGroup> replacedFileGroups, Pair<Boolean, List<CleanFileInfo>> expected) {
+    // setup savepoint mocks
+    Set<String> savepointTimestamps = savepoints.stream().map(Pair::getLeft).collect(Collectors.toSet());
+    when(mockHoodieTable.getSavepointTimestamps()).thenReturn(savepointTimestamps);
+    if (!savepoints.isEmpty()) {
+      HoodieActiveTimeline activeTimeline = mock(HoodieActiveTimeline.class);
+      when(mockHoodieTable.getActiveTimeline()).thenReturn(activeTimeline);
+      for (Pair<String, Option<byte[]>> savepoint : savepoints) {
+        HoodieInstant instant = new HoodieInstant(false, HoodieTimeline.SAVEPOINT_ACTION, savepoint.getLeft());
+        when(activeTimeline.getInstantDetails(instant)).thenReturn(savepoint.getRight());
+      }
+    }
+    String partitionPath = "partition1";
+    // setup replaced file groups mocks
+    when(mockFsView.getReplacedFileGroupsBefore(earliestInstant, partitionPath)).thenReturn(replacedFileGroups.stream());
+    // setup current file groups mocks
+    when(mockFsView.getAllFileGroupsStateless(partitionPath)).thenReturn(allFileGroups.stream());
+    HoodieWriteConfig config = HoodieWriteConfig.newBuilder().withPath("/tmp")
+        .withCleanConfig(HoodieCleanConfig.newBuilder()
+            .cleanerNumHoursRetained(24)
+            .withCleanerPolicy(HoodieCleaningPolicy.KEEP_LATEST_BY_HOURS)
+            .build())
+        .build();
+    CleanPlanner<?, ?, ?, ?> cleanPlanner = new CleanPlanner<>(context, mockHoodieTable, config);
+    HoodieInstant earliestCommitToRetain = new HoodieInstant(HoodieInstant.State.COMPLETED, "COMMIT", earliestInstant);
+    Pair<Boolean, List<CleanFileInfo>> actual = cleanPlanner.getDeletePaths(partitionPath, Option.of(earliestCommitToRetain));
+    assertEquals(expected, actual);
+  }
+
+  static Stream<Arguments> keepLatestByHoursArgs() {
```

Comment:
it would be easier for maintenance if the test scenario can be described in javadoc

Reply:
I think keeping it inline is easier actually since you can see the args and people will not forget to update the comments as they are in-line